presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestJoinNodeFlattener.java

```diff
@@ -311,10 +314,83 @@ public void testMoreThanJoinLimit()
                 .setFilter(and(createEqualsExpression(a1, c1), createEqualsExpression(b1, e1)))
                 .setOutputVariables(a1, b1, c1, d1, d2, e1, e2)
                 .build();
-        assertEquals(toMultiJoinNode(joinNode, noLookup(), 2, functionResolution, determinismEvaluator), expected);
+        assertEquals(toMultiJoinNode(joinNode, noLookup(), 2, true, functionResolution, determinismEvaluator), expected);
     }
 
-    private RowExpression createEqualsExpression(VariableReferenceExpression left, VariableReferenceExpression right)
+    @Test
+    public void testProjectNodesBetweenJoinNodesAreFlattenedForComplexEquiJoins()
+    {
+        PlanBuilder p = planBuilder();
+        VariableReferenceExpression a1 = p.variable("A1");
+        VariableReferenceExpression b1 = p.variable("B1");
+        VariableReferenceExpression c1 = p.variable("C1");
+        VariableReferenceExpression d1 = p.variable("D1");
+        VariableReferenceExpression e1 = p.variable("E1");
+        VariableReferenceExpression sum = p.variable("SUM");
+        VariableReferenceExpression rename = p.variable("RENAME");
+        VariableReferenceExpression renamePlusSum = p.variable("RENAME_PLUS_SUM");
+
+        ValuesNode valuesA = p.values(a1);
+        ValuesNode valuesB = p.values(b1);
+        ValuesNode valuesC = p.values(c1);
+        ValuesNode valuesD = p.values(d1);
+        ValuesNode valuesE = p.values(e1);
+        Assignments assignmentA1PlusB1 = Assignments.builder().put(sum, createAddExpression(a1, b1)).build();
+        Assignments assignmentRenameC1 = Assignments.builder().put(rename, c1).build();
+        Assignments assignmentRenamePlusSum = Assignments.builder().put(renamePlusSum, createAddExpression(rename, sum)).build();
+
+        ProjectNode projectOverJoin3 = p.project(assignmentRenamePlusSum, p.join(
+                INNER,
+                p.project(assignmentA1PlusB1, p.join(// projectOverJoin1
+                        INNER,
+                        valuesA,
+                        valuesB,
+                        ImmutableList.of(equiJoinClause(a1, b1)),
+                        ImmutableList.of(a1, b1),
+                        Optional.empty())),
+                p.project(assignmentRenameC1, p.join(// projectOverJoin2
+                        INNER,
+                        valuesC,
+                        valuesD,
+                        ImmutableList.of(equiJoinClause(c1, d1)),
+                        ImmutableList.of(c1),
+                        Optional.empty())),
+                ImmutableList.of(equiJoinClause(sum, rename)),
+                ImmutableList.of(sum, rename),
+                Optional.empty()));
+
+        JoinNode topMostJoinNode = p.join(
+                INNER,
+                valuesE,
+                projectOverJoin3,
+                ImmutableList.of(equiJoinClause(e1, renamePlusSum)),
+                ImmutableList.of(e1, renamePlusSum),
```

Comment:
`renamePlusSum` is a variable from an intermediate Project - this gets set on the overall assignment of the MultiJoinNode with the resolved variables of the input sources